netket/graph/lattice.py

```diff
@@ -64,99 +63,15 @@ def __repr__(self):
         return f"LatticeSite({s})"
 
 
-def create_sites(
-    basis_vectors, extent, apositions, pbc, order
-) -> Tuple[Sequence[LatticeSite], Sequence[bool], Dict[HashableArray, int]]:
-    # note: by modifying these, the number of shells can be tuned.
-    shell_vec = _np.where(pbc, 2 * order, 0)
-    shift_vec = _np.where(pbc, order, 0)
-
-    shell_min = 0 - shift_vec
-    shell_max = _np.asarray(extent) + shell_vec - shift_vec
-    # cell coordinates
-    ranges = [slice(lo, hi) for lo, hi in zip(shell_min, shell_max)]
-    # site coordinate within unit cell
-    ranges += [slice(0, len(apositions))]
-
-    basis_coords = _np.vstack([_np.ravel(x) for x in _np.mgrid[ranges]]).T
-    site_coords = (
-        basis_coords[:, :-1]
-        + _np.tile(apositions.T, reps=len(basis_coords) // len(apositions)).T
-    )
-    positions = site_coords @ basis_vectors
-
-    sites = []
-    coord_to_site = {}
-    for idx, (coord, pos) in enumerate(zip(basis_coords, positions)):
-        sites.append(
-            LatticeSite(
-                id=None,  # to be set later, after sorting all sites
-                basis_coord=coord,
-                position=pos,
-            ),
-        )
-        coord_to_site[HashableArray(coord)] = idx
-    is_inside = ~(
-        _np.any(basis_coords[:, :-1] < 0, axis=1)
-        | _np.any(basis_coords[:, :-1] > (extent - 1), axis=1)
-    )
-    return sites, is_inside, coord_to_site
+def _create_sites(basis_vectors, extent, site_offsets):
+    basis_coords, positions = create_site_positions(basis_vectors, extent, site_offsets)
 
+    sites = [
+        LatticeSite(id=idx, position=pos, basis_coord=coord)
+        for idx, (coord, pos) in enumerate(zip(basis_coords, positions))
+    ]
```

Comment:
I'm a bit lost on the  logic here...

Doesn't a single `LatticeSite` have the whole informations about `basis_coords` and `positions`? 
Why does `Lattice` store the duplicated information?

Reply:
Tbf I'm not sure why these `LatticeSite` objects exist, I just adapted the logic for making them to avoid breaking anything

Reply:
I agree they duplicate information

Reply:
@femtobit ?

Reply:
can't we remove the other stuff? basis_coords and positions?

Reply:
Those are arrays (and used as such in the code). `.sites` is a sequence of objects.